### PR TITLE
refactor: Generate field ids using a sequence

### LINKED ISSUE #2333: FieldDescription IDs are dependent on which fields are currently on the schema
This creates problems for schema branching and field removal.

For example, if two different fields are added in different schema branches, they will have the same ID (as the number of other fields is the same).  This means that data saved in documents on the two branches will share the same space/datastore-keys and do weird stuff.

WARNING: This cannot be solved simply by using a sequence for the IDs at the moment, as the ID field (local concept) is current held on the schema (global).  The fact that other fields have been added on different schema branches (incrementing the sequence) must not impact the schema version ids (via the field id).

### PR DESCRIPTION
**Pull request by AndrewSisley**
## Relevant issue(s)

Resolves #2333

## Description

Generates field ids using a sequence instead of their index in the array.

This is required for allowing field deletion and schema branching as otherwise those two will alter the position and thus id of fields.  For example, previously if two schema branches were to add a new field each, the new fields would share the same id and mess the datastore cache up in interesting ways.

## WARNING
This PR is currently branched off of https://github.com/sourcenetwork/defradb/pull/2336 - please only review commits from 'Simplify isNew evaluation' onwards here.

### CODE REVIEW

File: client/index.go [hunk 1]
```diff
@@ -41,18 +41,19 @@ type IndexDescription struct {
...
-			for i := range schema.Fields {
-				colField := schema.Fields[i]
-				if field.Name == colField.Name && !fieldsMap[field.Name] {
-					fieldsMap[field.Name] = true
-					fields = append(fields, colField)
-					break
-				}
+			if fieldsMap[field.Name] {
+				// If the FieldDescription has already been added to the result do not add it a second time
+				// this can happen if a field if referenced by multiple indexes
```
Comment (islamaliev) [inline client/index.go:51]: typo: "if a field referenced"

File: planner/mapper/mapper.go [hunk 2]
```diff
@@ -267,7 +267,7 @@ func resolveAggregates(
 	inputFields []Requestable,
 	mapping *core.DocumentMapping,
 	collectionName string,
-	schema client.SchemaDescription,
+	schema client.CollectionDefinition,
```
Comment (islamaliev) [inline planner/mapper/mapper.go:270]: todo: rename the param to `colDef` or something similar please

**Review by islamaliev — COMMENTED** [review]
Overall looks great. I just have few comments.

File: client/definitions.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+		}
+	}
+	return fields
+}
+
+// FieldDefinition describes the combined local and global set of properties that constitutes
+// a field on a collection.
+//
+// It draws it's information from the [CollectionFieldDescription] on the [CollectionDescription],
+// and the [SchemaFieldDescription] on the [SchemaDescription].
```
[the remark this answers is not in the document]
Comment (AndrewSisley) [reply client/definitions.go:57]: Moving to https://github.com/sourcenetwork/defradb/pull/2336

File: client/definitions.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+//
+// It is to [CollectionFieldDescription] and [SchemaFieldDescription] what [CollectionDefinition]
+// is to [CollectionDescription] and [SchemaDescription].
+type FieldDefinition struct {
```
[the remark this answers is not in the document]
Comment (AndrewSisley) [reply client/definitions.go:61]: I do very much believe copy-paste here is much cheaper than detangling inheritance hierarchies.  There has been a bit of a discussion already on this in the PR here: https://github.com/sourcenetwork/defradb/pull/2336#discussion_r1501059384

I'm pasting your comment over there and closing this one, please continue the discussion over there :)

File: client/index.go (hunk 1 above)
Comment (AndrewSisley) [reply client/index.go:51]: Moved to https://github.com/sourcenetwork/defradb/pull/2336/files#r1504574665

File: planner/mapper/mapper.go (hunk 2 above)
Comment (AndrewSisley) [reply planner/mapper/mapper.go:270]: Moved to https://github.com/sourcenetwork/defradb/pull/2336/files#r1504578581

File: core/key.go [hunk 5]
```diff
@@ -165,11 +167,29 @@ type P2PCollectionKey struct {
...
+type IndexIDSequenceKey struct {
+	CollectionID uint32
 }
 
-var _ Key = (*SequenceKey)(nil)
+var _ Key = (*IndexIDSequenceKey)(nil)
+
+// FieldIDSequenceKey is used to key the sequence used to generate field ids.
+//
+// The sequence is specific to each collection root.  Multiple collection of the same root
```
Comment (fredcarle) [inline core/key.go:186]: question: Can you remind me what the root is? Do you mean the root schema?

Comment (fredcarle) [reply core/key.go:186]: Nevermind. I saw the answer bellow 😅

**Review by fredcarle — APPROVED** [review]
LGTM!

**Pull request merged**